openshift/catalog-sync/pkg/extract/extract.go

```diff
@@ -0,0 +1,180 @@
+	policy := &signature.Policy{
+		Default: []signature.PolicyRequirement{
+			signature.NewPRInsecureAcceptAnything(),
+		},
+	}
+	policyCtx, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return fmt.Errorf("create policy context: %w", err)
+	}
+	defer policyCtx.Destroy()
+
+	// Force image resolution to Linux platform to avoid OS mismatch errors
+	// during testing on Mac. Example error:
+	// err: <*errors.errorString>{
+	//     s: "no image found in manifest list for architecture \"arm64\", variant \"v8\", OS \"darwin\"",
+	// }
+	sysCtx := &types.SystemContext{
+		OSChoice: "linux",
+	}
+
+	if _, err := imgcopy.Image(ctx, policyCtx, destRef, srcRef, &imgcopy.Options{
+		SourceCtx:      sysCtx,
+		DestinationCtx: sysCtx,
+	}); err != nil {
+		return fmt.Errorf("copy image: %w", err)
+	}
+
+	return nil
+}
+
+// extractFilesystem extracts the filesystem layers from the OCI image into <tmpDir>/fs
+func extractFilesystem(ctx context.Context, ociDir string) error {
+	ref, err := layout.ParseReference(fmt.Sprintf("%s:%s", ociDir, "v1"))
+	if err != nil {
+		return fmt.Errorf("parse layout: %w", err)
+	}
+	src, err := ref.NewImageSource(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("open image source: %w", err)
+	}
+	defer src.Close()
+
+	manifests, _, err := src.GetManifest(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("get manifest: %w", err)
+	}
+
+	mf, err := manifest.FromBlob(manifests, manifest.GuessMIMEType(manifests))
+	if err != nil {
+		return fmt.Errorf("parse manifest: %w", err)
+	}
+
+	dir := filepath.Join(ociDir, "fs")
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("make fs dir: %w", err)
+	}
+
+	for i, layer := range mf.LayerInfos() {
+		rc, _, err := src.GetBlob(ctx, layer.BlobInfo, nil)
+		if err != nil {
+			return fmt.Errorf("get blob %d: %w", i, err)
+		}
+
+		decompressed, _, err := compression.AutoDecompress(rc)
+		if err != nil {
+			rc.Close()
+			return fmt.Errorf("decompress blob %d: %w", i, err)
+		}
+
+		// To avoid permission errors faced
+		mask := os.FileMode(0755)
+		opts := &archive.TarOptions{
+			NoLchown:          true,
+			IgnoreChownErrors: true,
+			InUserNS:          true,
+			ForceMask:         &mask,
+		}
+
+		_, err = archive.ApplyUncompressedLayer(dir, decompressed, opts)
+		decompressed.Close()
+		rc.Close()
+
+		if err != nil {
+			if isPermError(err) {
+				fmt.Fprintf(os.Stderr, "[WARNING] skipping layer %d: %v\n", i, err)
+				continue
+			}
+			return fmt.Errorf("apply layer %d: %w", i, err)
+		}
+	}
+	return nil
+}
+
+// isPermError detects known non-critical permission errors
+// required to Linux/Mac compatibility
+func isPermError(err error) bool {
+	msg := err.Error()
+	return strings.Contains(msg, "permission denied") &&
+		(strings.Contains(msg, ".bash_logout") ||
+			strings.Contains(msg, ".pwd.lock") ||
+			strings.Contains(msg, "/afs") ||
+			strings.Contains(msg, "/etc/X11"))
```

Comment:
This seems fragile. If a single file in a layer has a problem like this, it seems like the rest of that layer's contents will be skipped, which isn't great. I think I saw something in `containers/storage/pkg/archive` that wraps a tar reader that allows the wrapper to make arbitrary changes to the tar header and to the contents of the files, and also allowing the entire entry to be skipped.

Ideally, we handle all files, directories, symlinks, and hardlinks with the necessary chown/chmod/unset xattrs, and then error if we encounter any other type (e.g. socket, block device, pipe, character device, etc.)

Reply:
I tried to use the options from archive but I will give look.

Reply:
I could clean it all seems that the issues that I faced before are no longer more faced anymore
Thx.